### PR TITLE
Coding Style: Add guidelines for local functions

### PR DESCRIPTION
**Pull request by mary-georgiou-sonarsource**

### CODE REVIEW

File: docs/coding-style.md [hunk 1]
```diff
@@ -39,6 +39,14 @@ int CallerTwo() => Leaf() + PublicMethod();
 int Leaf() =>  42;
 ```
 
+#### Local functions
```
Comment (pavel-mikula-sonarsource) [inline docs/coding-style.md:42]: It's weird to have 4th level `####` label under two level `## Class Member`

File: docs/coding-style.md [hunk 2]
```diff
@@ -39,6 +39,14 @@ int CallerTwo() => Leaf() + PublicMethod();
...
+
+There are no strict rules on when to use local functions. It should be decided on a case-by-case basis.
+
+By default, you should prefer methods over local functions. Use local functions if it makes the code significantly easier to understand. For example:
+- Accessing the method's local state directly, instead of using parameters, reduces noise.
+- The name of the function would not make sense at the class level.
+
 Local functions should always be placed at the end of a method.
```
Comment (pavel-mikula-sonarsource) [inline docs/coding-style.md:50]: While you're on it, you can fix the syntax error on line 57, `return` is not valid with arrow method

File: docs/coding-style.md [hunk 3]
```diff
@@ -39,6 +39,14 @@ int CallerTwo() => Leaf() + PublicMethod();
 int Leaf() =>  42;
 ```
 
+#### Local functions
+
+There are no strict rules on when to use local functions. It should be decided on a case-by-case basis.
```
Comment (zsolt-kolbay-sonarsource) [inline docs/coding-style.md:44]: Please fix the indentation of the code sample in line 53.

**Review by zsolt-kolbay-sonarsource — APPROVED** [review]
LGTM. Left one more comment.

**Pull request merged**